### PR TITLE
fix: replacing setters to ngonchanges

### PR DESCRIPTION
**Pull request by rengare**
#### Please provide a brief summary of this pull request.
Replaced setters in favour of ngOnChanges for building component CSS classes.

[Code-Contribution-Guideline](https://github.com/SAP/fundamental-ngx/wiki/Code-Contribution-Guideline#building-component-css-classes) has been updated

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done

### CODE REVIEW

File: libs/core/src/lib/bar/bar.component.ts [hunk 1]
```diff
@@ -17,95 +17,66 @@ export type BarDesignType = 'header' | 'subheader' | 'header-with-subheader' | '
...
-
-    /** @hidden */
-    elementRef(): ElementRef<any> {
+    public elementRef(): ElementRef<any> {
         return this._elementRef;
     }
 
     /** @hidden */
-    ngOnChanges(): void {
+    public ngOnChanges(): void {
```
Comment (InnaAtanasova) [inline libs/core/src/lib/bar/bar.component.ts]: Aren't these public by default?

Comment (rengare) [reply libs/core/src/lib/bar/bar.component.ts]: public access modifier removed

**Comment by droshev** [comment]
@rengare Travis and netlify are failing.

**Comment by mikerodonnell89** [comment]
The Tokenizer pattern requires us to use both fd-form-control and fd-tokenizer-input directives on the same input element.  I tried having both fd-form-control and fd-tokenizer-input directives implement CssClassBuilder, and when I do this, only the classes from one directive will be applied to the element, so I end up with `class="fd-tokenizer__input` when I want `class="fd-tokenizer__input fd-input fd-input--compact`.

**Comment by rengare** [comment]
> The Tokenizer pattern requires us to use both fd-form-control and fd-tokenizer-input directives on the same input element. I tried having both fd-form-control and fd-tokenizer-input directives implement CssClassBuilder, and when I do this, only the classes from one directive will be applied to the element, so I end up with `class="fd-tokenizer__input` when I want `class="fd-tokenizer__input fd-input fd-input--compact`.

I'll check that

File: libs/core/src/lib/tabs/tab-utils/tab-directives.ts [hunk 2]
```diff
@@ -81,23 +82,20 @@ export class TabIconDirective implements CssClassBuilder, OnInit {
...
-    ) {}
+    ) { }
 
     /** @hidden
      * Function runs when component is initialized
      * function should build component css class
      * function should build css style
      */
-    ngOnInit(): void {
+    public ngOnChanges(): void {
```
Comment (InnaAtanasova) [inline libs/core/src/lib/tabs/tab-utils/tab-directives.ts:98]: Don't hate me :D You need to remove the `public` :)

File: libs/core/src/lib/tabs/tab-utils/tab-directives.ts [hunk 3]
```diff
@@ -106,18 +104,18 @@ export class TabIconDirective implements CssClassBuilder, OnInit {
      * function must return single string
      * function is responsible for order which css classes are applied
      */
-    buildComponentCssClass(): string {
+    public buildComponentCssClass(): string {
```
Comment (InnaAtanasova) [inline libs/core/src/lib/tabs/tab-utils/tab-directives.ts:107]: `public`

File: libs/core/src/lib/tabs/tab-utils/tab-directives.ts [hunk 4]
```diff
@@ -106,18 +104,18 @@ export class TabIconDirective implements CssClassBuilder, OnInit {
...
+            this.icon ? `sap-icon--${this.icon}` : '',
+            this.class
         ].filter(x => x !== '').join(' ');
     }
 
     /** HasElementRef interface implementation
      * function used by applyCssClass and applyCssStyle decorators
      */
-    elementRef(): ElementRef<any> {
+    public elementRef(): ElementRef<any> {
```
Comment (InnaAtanasova) [inline libs/core/src/lib/tabs/tab-utils/tab-directives.ts:118]: `public`

**Comment by InnaAtanasova** [comment]
> Can we also use this approach and remove the AbstractFdNgxClass?

I think this is outside of the scope of this PR. This PR was supposed to get rid of the private Input pros and setters/getters we recently introduced. We agreed that once we start refactoring a component we will migrate the `AbstractFdNgxClass` to this approach. At some point we can create an issue for this and go over all components and change the ones that are not migrated.

**Comment by InnaAtanasova** [comment]
@mikerodonnell89 , could you take a look at the PR again and add a new review? Thanks

File: libs/core/src/lib/tabs/tab-utils/tab-directives.ts (hunk 2 above)
Comment (rengare) [reply libs/core/src/lib/tabs/tab-utils/tab-directives.ts:98]: :) good point, I'll do that

File: libs/core/src/lib/tabs/tab-utils/tab-directives.ts (hunk 4 above)
Comment (rengare) [reply libs/core/src/lib/tabs/tab-utils/tab-directives.ts:118]: removed

File: libs/core/src/lib/tabs/tab-utils/tab-directives.ts (hunk 3 above)
Comment (rengare) [reply libs/core/src/lib/tabs/tab-utils/tab-directives.ts:107]: removed

File: libs/core/src/lib/tabs/tab-utils/tab-directives.ts (hunk 2 above)
Comment (rengare) [reply libs/core/src/lib/tabs/tab-utils/tab-directives.ts:98]: removed

File: libs/core/src/lib/tabs/tab-utils/tab-directives.ts [hunk 5]
```diff
@@ -126,70 +125,61 @@ export class TabIconDirective implements CssClassBuilder, OnInit {
     // tslint:disable-next-line:directive-selector
     selector: '[fd-tab-tag]',
     host: {
-        'class': 'fd-tabs__tag'
-    }
+        class: 'fd-tabs__tag',
+    },
```
Comment (JKMarkowski) [inline libs/core/src/lib/tabs/tab-utils/tab-directives.ts:129]: IT's really detail, but I see that in `tab-directives` there are a lot of hanging commas. Could you remove them ?

**Comment by mikerodonnell89** [comment]
This plus https://github.com/SAP/fundamental-ngx/pull/2249 looks great

File: libs/core/src/lib/tabs/tab-utils/tab-directives.ts (hunk 5 above)
Comment (rengare) [reply libs/core/src/lib/tabs/tab-utils/tab-directives.ts]: removed

**Pull request merged**